### PR TITLE
Refactor _utilities-spacing.scss

### LINKED ISSUE #18129: Refactor _utilities-spacing.scss using loops to decrease redundancy

### PR DESCRIPTION
**Pull request by cvrebert**
Fixes #18129.

For uniformity, this also adds a `-sm` suffix to the classes which formerly has no suffix.
CC: @mdo for review

### CODE REVIEW

**Comment by michaseel** [comment]
Look at #18128: i proposed to use multiplier numbers instead of shirt sizes for spacing utility class names

**Comment by cvrebert** [comment]
The exact naming is basically an orthogonal issue. If anything, this refactoring makes it easier to apply such a renaming.

**Comment by mdo** [comment]
I'd rather we didn't use `sm` (and the other sizes even tbh) since it's easily confused with the grid tier sizes. I'd imagine we'd get a ton of issues around that.

**Comment by cvrebert** [comment]
@mdo SGTM. What size naming scheme do you want to use instead?

**Comment by mdo** [comment]
Unsure really. I've seen other folks use an array like `0`-`3` (e.g., `.m-a-3`), but that's not very clear to me either.

**Comment by cvrebert** [comment]
@mdo Well, this PR just refactors the existing code and doesn't really change the naming scheme. Could we merge the refactoring and leave coming up with a better naming scheme to #18128?

File: scss/_utilities-spacing.scss [hunk 1]
```diff
@@ -1,71 +1,33 @@
...
-.p-l-lg { padding-left:   ($spacer-y * 3) !important; }
-.p-x-lg { padding-right:  ($spacer-x * 3) !important; padding-left:   ($spacer-x * 3) !important; }
-.p-y-lg { padding-top:    ($spacer-y * 3) !important; padding-bottom: ($spacer-y * 3) !important; }
+@each $prop, $abbrev in (margin: m, padding: p) {
+  @each $size, $lengths in $spacers {
+    $length-x:   map-get($lengths, x);
+    $length-y:   map-get($lengths, y);
+    $length-all: map-get($lengths, all);
+
+    .#{$abbrev}-a-#{$size} { #{$prop}:        $length-all !important; } // a = All sides
```
Comment (cvrebert) [inline scss/_utilities-spacing.scss]: A user argues in #18339 that this should instead be
```
#{$prop}: $length-y $length-x;
```

Comment (cvrebert) [inline scss/_utilities-spacing.scss]: @mdo Thoughts on this suggestion?

Comment (mdo) [inline scss/_utilities-spacing.scss]: Is there a difference? Looks like that might only save on a line of code by eliminating `$length-all`? Or is it the ordering thing from that issue?

**Comment by mdo** [comment]
Looking at this again, I'd like us to use a numerical scale (which just so happens to build nicely upon the `.*-*-0` classes we have.

File: scss/_utilities-spacing.scss (hunk 1 above)
Comment (cvrebert) [inline scss/_utilities-spacing.scss]: @mdo 
> Is there a difference?

Yes. If `$length-all` ≠ `$length-x` OR `$length-all` ≠ `$length-y`, then the result is different.
`$length-all` on all sides vs. `$length-x` horizontally & `$length-y` vertically

**Comment by cvrebert** [comment]
@mdo So, like `.*-*-1`, `.*-*-2`, `.*-*-3`; or did you have something else in mind?

**Comment by mdo** [comment]
Yup, that works I think.

**Comment by cvrebert** [comment]
Okay, switched to a numerical naming scheme.
That just leaves the `.*-a-*`definition issue.

File: scss/_utilities-spacing.scss (hunk 1 above)
Comment (mdo) [inline scss/_utilities-spacing.scss]: Shit man I dunno :D. Your call.

Comment (cvrebert) [inline scss/_utilities-spacing.scss]: I'll go with the user's suggestion then. It also has the benefit of killing `$*-all`.

**Comment by cvrebert** [comment]
Definition issue resolved:
```scss
.#{$abbrev}-a-#{$size} { #{$prop}:        $length-y $length-x !important; }
```

**Comment by sgb-io** [comment]
For anyone who got tripped up using an earlier version of the alpha and needs to migrate, this is how the shirtsizes map now.

| before              | after              |   |   |   |
|---------------------|--------------------|---|---|---|
| `.{prop}-{side}-md` | `.{prop}-{side}-2` |   |   |   |
| `.{prop}-{side}-lg` | `.{prop}-{side}-3` |   |   |   |
|                     |                    |   |   |   |

**Pull request merged**